src/jitlog/generator.lua

```diff
@@ -339,7 +333,431 @@ function parser:complete()
+  local vtotal_template
+  local tmpldata = {
+    name = logfunc_getfieldvar(msgdef, f), 
+    sizename = f.name.."_size",
+  }
+  
+  local vtype = self.types[f.type]
+  tmpldata.element_size = vtype.element_size or vtype.size
+  
+  -- Check that the length is not an implicit arg after the field
+  if f.buflen then
+    tmpldata.sizename = f.buflen
+    local szfield = msgdef.fieldlookup[f.buflen]
+    if szfield then
+      tmpldata.sizename = logfunc_getfieldvar(msgdef, szfield)
+    end
+  end
+
+  if f.type == "string" and (f.implicitlen or msgdef.use_msgsize == f.name) then
+    vtotal_template = "  MSize {{sizename}} = (MSize)strlen({{name}})+1;\n  vtotal += {{sizename}};\n"
+  else
+    vtotal_template = "  vtotal += {{sizename}} * {{element_size}};\n"
+  end
+  vtotal = vtotal .. self:buildtemplate(vtotal_template, tmpldata)
+  vwrite = vwrite .. self:buildtemplate("  lj_buf_putmem(sb, {{name}}, (MSize)({{sizename}} * {{element_size}}));\n", tmpldata)
+
+  return vtotal, vwrite
+end
+
+local funcdef_fixed = [[
+LJ_STATIC_ASSERT(sizeof(MSG_{{name}}) == {{msgsize}});
+
+static LJ_AINLINE void log_{{name}}({{args}})
+{
+  SBuf *sb = (SBuf *)g->vmevent_data;
+  MSG_{{name}} *msg = (MSG_{{name}} *)sbufP(sb);
+{{fields:  %s\n}}  setsbufP(sb, sbufP(sb) + {{msgsize}});
+  lj_buf_more(sb, {{minbuffspace}});
+}
+
+]]
+
+local funcdef_vsize = [[
+LJ_STATIC_ASSERT(sizeof(MSG_{{name}}) == {{msgsize}});
+
+static LJ_AINLINE void log_{{name}}({{args}})
+{
+  SBuf *sb = (SBuf *)g->vmevent_data;
+  MSG_{{name}} *msg;
+{{vtotal}}  msg = (MSG_{{name}} *)lj_buf_more(sb, (MSize)(vtotal + {{minbuffspace}}));
+{{fields:  %s\n}}  setsbufP(sb, sbufP(sb) + {{msgsize}});
+{{vwrite}}
+}
+
+]]
+
+generator.custom_field_writers = {
+  timestamp_highres = "__rdtsc();",
+  timestamp = "__rdtsc();",
+  gettime = "__rdtsc();",
+  setref = function(self, msgdef, f, valuestr)
+    local setref = (f.type == "MRef" and "setmref") or "setgcrefp"
+    local type = self.types[f.type]
+    if f.ptrarg or type.ptrarg or f.struct_addr then
+      return format("%s(msg->%s, %s);", setref, f.name, valuestr)
+    else
+      -- Just do an assignment for raw GCref values
+      return format("msg->%s = %s;", f.name, valuestr) 
+    end
+  end,
+  msgid = function(self, msgdef, f) 
+    return format("msg->msgid = MSGID_%s;", msgdef.name) 
+  end,
+  vtotal = "(uint32_t)vtotal;",
+  widenptr = function(self, msgdef, f, valuestr) 
+    return format("msg->%s = (uint64_t)(uintptr_t)(%s);", f.name, valuestr)
+  end,
+}
+
+function generator:write_logfunc(def)
+  local msgid = def.msgid
+  if not msgid then
+    local base = def.base and self.msglookup[def.base]
+    assert(not def.base or base, "missing base message base")
+    if base then
+      msgid = base.msgid or base.name
+    else
+      msgid = def.name
+    end
+  end
+   
+  local fields = {}
+  local vtotal, vwrite = "", ""
+  if def.vsize then
+    vtotal = format("  size_t vtotal = sizeof(MSG_%s);\n", def.name)
```

Comment:
Is the double space at the beginning of the string intentional?

Reply:
Its just indenting produced output but the template system can be passed a table and apply a string format to each value in it before concatenating them on it avoiding manually adding it to every line

Reply:
Is that a yes or a no? ;)

Reply:
its a yes with the caveat that there might be a better way to do it

Reply:
Is the better way fairly easy to do?

Reply:
I tried it out i think made things a bit cleaner at least